src/Dictionaries/RegExpTreeDictionary.cpp

```diff
@@ -0,0 +1,545 @@
+    }
+    std::set<UInt64> visited;
+    UInt64 topology_id = 0;
+    for (const auto & [id, value]: regex_nodes)
+        if (value->parent_id == 0) // this is root node.
+            initTopologyOrder(id, visited, topology_id);
+    if (topology_order.size() != regex_nodes.size())
+        throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Invalid Regex tree");
+}
+
+void RegExpTreeDictionary::initTopologyOrder(UInt64 node_idx, std::set<UInt64> & visited, UInt64 & topology_id)
+{
+    visited.insert(node_idx);
+    for (UInt64 child_idx : regex_nodes[node_idx]->children)
+        if (visited.contains(child_idx))
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Invalid Regex tree");
+        else
+            initTopologyOrder(child_idx, visited, topology_id);
+    topology_order[node_idx] = topology_id++;
+}
+
+void RegExpTreeDictionary::loadData()
+{
+    if (!source_ptr->hasUpdateField())
+    {
+        QueryPipeline pipeline(source_ptr->loadAll());
+        PullingPipelineExecutor executor(pipeline);
+
+        Block block;
+        while (executor.pull(block))
+        {
+            initRegexNodes(block);
+        }
+        initGraph();
+    }
+    else
+    {
+        throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Dictionary {} does not support updating manual fields", name);
+    }
+}
+
+RegExpTreeDictionary::RegExpTreeDictionary(
+    const StorageID & id_, const DictionaryStructure & structure_, DictionarySourcePtr source_ptr_, Configuration configuration_)
+    : IDictionary(id_), structure(structure_), source_ptr(source_ptr_), configuration(configuration_), logger(&Poco::Logger::get("RegExpTreeDictionary"))
+{
+    if (auto * ch_source = typeid_cast<ClickHouseDictionarySource *>(source_ptr.get()))
+    {
+        Block sample_block;
+        /// id, parent_id, regex, keys, values
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeUInt64()), kId));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeUInt64()), kParentId));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeString()), kRegExp));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeArray(DataTypePtr(new DataTypeString()))), kKeys));
+        sample_block.insert(ColumnWithTypeAndName(DataTypePtr(new DataTypeArray(DataTypePtr(new DataTypeString()))), kValues));
+        ch_source->sample_block = std::move(sample_block);
+    }
+
+    loadData();
+    calculateBytesAllocated();
+}
+
+String processBackRefs(const String & data, const re2_st::RE2 & searcher, const std::vector<StringPiece> & pieces)
+{
+    re2_st::StringPiece haystack(data.data(), data.size());
+    re2_st::StringPiece matches[10];
+    String result;
+    searcher.Match(haystack, 0, data.size(), re2_st::RE2::Anchor::UNANCHORED, matches, 10);
+    for (const auto & item : pieces)
+    {
+        if (item.ref_num >= 0)
+            result += matches[item.ref_num].ToString();
```

Comment:
We already checked `ref_num` in `createStringPieces`, but let's also add check here to be sure that `item.ref_num < 10` and we don't overflow `matches` array